cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/TimePartitionedStreamFileWriter.java

```diff
@@ -152,17 +157,37 @@ public int hashCode() {
 
       // Create the event and index file with the next sequence id
       int fileSequence = StreamUtils.getNextSequenceId(partitionDirectory, fileNamePrefix);
-      Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.EVENT);
-      Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
-                                                            fileSequence, StreamFileType.INDEX);
+      final Location eventFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.EVENT);
+      final Location indexFile = StreamUtils.createStreamLocation(partitionDirectory, fileNamePrefix,
+                                                                  fileSequence, StreamFileType.INDEX);
       // The creation should succeed, as it's expected to only have one process running per fileNamePrefix.
       if (!eventFile.createNew() || !indexFile.createNew()) {
         throw new IOException("Failed to create new file at " + eventFile + " and " + indexFile);
       }
 
       LOG.debug("New stream file created at {}", eventFile);
-      return new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      final StreamDataFileWriter streamDataFileWriter =
+        new StreamDataFileWriter(createOutputSupplier(eventFile), createOutputSupplier(indexFile), indexInterval);
+
+      timer.schedule(new TimerTask() {
```

Comment:
Instead, the `closePartitionWriter` should be used. You simply check against the current `timePartition`.